File: Data Structures/5. Swift/i. Sets.swift
struct CustomSet<T: Hashable> {
    private var elements: [T: Bool]  // Dictionary to mimic a set

    // MARK: - Basic Set Operations

    /// Check if the set is empty.
    func isEmpty() -> Bool {
        return elements.isEmpty
    }

    /// Get the number of elements in the set.
    func count() -> Int {
        return elements.count
    }

    /// Insert an element into the set.
    func insert(_ element: T) {
        elements[element] = true
    }

    /// Remove an element from the set.
    @discardableResult
    func remove(_ element: T) -> T? {
        return elements.removeValue(forKey: element) != nil ? element : nil
    }

    /// Check if the set contains a specific element.
    func contains(_ element: T) -> Bool {
        return elements[element] != nil
    }

    // MARK: - Set Operations

    /// Perform the union operation with another set and return a new set.
    func union(with otherSet: CustomSet<T>) -> CustomSet<T> {
        var newSet = self
        for element in otherSet.elements.keys {
            newSet.insert(element)
        }
        return newSet
    }

    /// Perform the intersection operation with another set and return a new set.
    func intersection(with otherSet: CustomSet<T>) -> CustomSet<T> {
        var newSet = CustomSet<T>()
        for element in elements.keys {
            if otherSet.contains(element) {
                newSet.insert(element)
            }
        }
        return newSet
    }

    /// Perform the difference operation with another set and return a new set.
    func difference(with otherSet: CustomSet<T>) -> CustomSet<T> {
        var newSet = self
        for element in otherSet.elements.keys {
            newSet.remove(element)
        }
        return newSet
    }

    /// Check if this set is a subset of another set.
    func isSubset(of otherSet: CustomSet<T>) -> Bool {
        for element in elements.keys {
            if !otherSet.contains(element) {
                return false
            }
        }
        return true
    }

    /// Check if this set is a superset of another set.
    func isSuperset(of otherSet: CustomSet<T>) -> Bool {
        return otherSet.isSubset(of: self)
    }

    /// Check if this set is disjoint with another set.
    func isDisjoint(with otherSet: CustomSet<T>) -> Bool {
        for element in elements.keys {
            if otherSet.contains(element) {
                return false
            }
        }
        return true
    }

    /// Apply a closure to each element in the set.
    func forEach(_ closure: (T) -> Void) {
        for element in elements.keys {
            closure(element)
        }
    }

    /// Remove all elements from the set.
    mutating func removeAll() {
        elements.removeAll()
    }

    /// Remove all occurrences of an element from the set.
    mutating func removeAllOccurrences(of element: T) {
        elements[element] = nil
    }

    /// Get the set with common elements between this set and another set.
    func symmetricDifference(with otherSet: CustomSet<T>) -> CustomSet<T> {
        let diff1 = self.difference(with: otherSet)
        let diff2 = otherSet.difference(with: self)
        return diff1.union(with: diff2)
    }

    /// Insert elements from another set into this set.
    mutating func formUnion(with otherSet: CustomSet<T>) {
        self = self.union(with: otherSet)
    }

    /// Remove elements of this set that are not in another set.
    mutating func formIntersection(with otherSet: CustomSet<T>) {
        self = self.intersection(with: otherSet)
    }

    /// Remove elements of this set that are in another set.
    mutating func subtract(_ otherSet: CustomSet<T>) {
        self = self.difference(with: otherSet)
    }

    // ... Add more set operations as needed ...
}

File: data_structures/1_cpp/i_sets.cpp
#include <unordered_map>
#include <iostream>

template <typename T>
class CustomSet {
private:
    std::unordered_map<T, bool> elements;  // Hash table to mimic a set

public:
    // Basic Set Operations

    bool isEmpty() const {
        return elements.empty();
    }

    int count() const {
        return elements.size();
    }

    void insert(const T& element) {
        elements[element] = true;
    }

    bool remove(const T& element) {
        return elements.erase(element) > 0;
    }

    bool contains(const T& element) const {
        return elements.find(element) != elements.end();
    }

    // Set Operations

    CustomSet<T> unionWith(const CustomSet<T>& otherSet) const {
        CustomSet<T> newSet(*this);
        for (const auto& entry : otherSet.elements) {
            newSet.insert(entry.first);
        }
        return newSet;
    }

    CustomSet<T> intersectionWith(const CustomSet<T>& otherSet) const {
        CustomSet<T> newSet;
        for (const auto& entry : elements) {
            if (otherSet.contains(entry.first)) {
                newSet.insert(entry.first);
            }
        }
        return newSet;
    }

    CustomSet<T> differenceWith(const CustomSet<T>& otherSet) const {
        CustomSet<T> newSet(*this);
        for (const auto& entry : otherSet.elements) {
            newSet.remove(entry.first);
        }
        return newSet;
    }

    bool isSubsetOf(const CustomSet<T>& otherSet) const {
        for (const auto& entry : elements) {
            if (!otherSet.contains(entry.first)) {
                return false;
            }
        }
        return true;
    }

    bool isSupersetOf(const CustomSet<T>& otherSet) const {
        return otherSet.isSubsetOf(*this);
    }

    bool isDisjointWith(const CustomSet<T>& otherSet) const {
        for (const auto& entry : elements) {
            if (otherSet.contains(entry.first)) {
                return false;
            }
        }
        return true;
    }

    void forEach(void (*closure)(const T&)) const {
        for (const auto& entry : elements) {
            closure(entry.first);
        }
    }

    void removeAll() {
        elements.clear();
    }

    void removeAllOccurrencesOf(const T& element) {
        elements.erase(element);
    }

    CustomSet<T> symmetricDifferenceWith(const CustomSet<T>& otherSet) const {
        CustomSet<T> diff1 = this->differenceWith(otherSet);
        CustomSet<T> diff2 = otherSet.differenceWith(*this);
        return diff1.unionWith(diff2);
    }

    void formUnionWith(const CustomSet<T>& otherSet) {
        *this = this->unionWith(otherSet);
    }

    void formIntersectionWith(const CustomSet<T>& otherSet) {
        *this = this->intersectionWith(otherSet);
    }

    void subtract(const CustomSet<T>& otherSet) {
        *this = this->differenceWith(otherSet);
    }

    // ... Add more set operations as needed ...
};

// Example usage
int main() {
    CustomSet<int> set1;
    set1.insert(1);
    set1.insert(2);
    set1.insert(3);

    CustomSet<int> set2;
    set2.insert(2);
    set2.insert(3);
    set2.insert(4);

    CustomSet<int> unionSet = set1.unionWith(set2);
    std::cout << "Union set: ";
    unionSet.forEach([](const int& element) {
        std::cout << element << " ";
    });
    std::cout << std::endl;

    CustomSet<int> intersectionSet = set1.intersectionWith(set2);
    std::cout << "Intersection set: ";
    intersectionSet.forEach([](const int& element) {
        std::cout << element << " ";
    });
    std::cout << std::endl;

    return 0;
}
